src/lib.rs

```diff
@@ -1,287 +1,284 @@
-use fvm_ipld_encoding::{RawBytes, DAG_CBOR};
-use fvm_sdk as sdk;
-use fvm_sdk::NO_DATA_BLOCK_ID;
-use fvm_shared::actor::builtin::Type;
-use fvm_shared::address::Address;
-use fvm_shared::bigint::bigint_ser::BigIntDe;
+use fil_actors_runtime::runtime::{ActorCode, Runtime};
+use fil_actors_runtime::{actor_error, cbor, ActorDowncast, ActorError, INIT_ACTOR_ADDR};
+use fvm_ipld_blockstore::Blockstore;
+use fvm_ipld_encoding::RawBytes;
+
 use fvm_shared::econ::TokenAmount;
-use fvm_shared::{ActorID, METHOD_SEND};
-use num_traits::Zero;
-use sdk::actor::get_actor_code_cid;
-use state::get_votes;
+use fvm_shared::error::ExitCode;
+use fvm_shared::{MethodNum, METHOD_CONSTRUCTOR};
+use ipc_gateway::MIN_COLLATERAL_AMOUNT;
+use num_derive::FromPrimitive;
+use num_traits::{FromPrimitive, Zero};
 
-use crate::blockstore::*;
-use crate::state::{get_stake, State};
+use crate::state::State;
 use crate::types::*;
 use crate::utils::*;
 
-pub const TEST_ADDR_ID: ActorID = 339;
-
-/// The actor's WASM entrypoint. It takes the ID of the parameters block,
-/// and returns the ID of the return value block, or NO_DATA_BLOCK_ID if no
-/// return value.
-///
-/// Should probably have macros similar to the ones on fvm.filecoin.io snippets.
-/// Put all methods inside an impl struct and annotate it with a derive macro
-/// that handles state serde and dispatch.
-#[no_mangle]
-pub fn invoke(params: u32) -> u32 {
-    let params = sdk::message::params_raw(params).unwrap().1;
-    let params = RawBytes::new(params);
-    // Conduct method dispatch. Handle input parameters and return data.
-    let ret: anyhow::Result<Option<RawBytes>> = match sdk::message::method_number() {
-        1 => Actor::constructor(deserialize_params(&params).unwrap()),
-        2 => Actor::join(deserialize_params(&params).unwrap()),
-        3 => Actor::leave(),
-        4 => Actor::kill(),
-        5 => Actor::submit_checkpoint(deserialize_params(&params).unwrap()),
-        _ => abort!(USR_UNHANDLED_MESSAGE, "unrecognized method"),
-    };
+fil_actors_runtime::wasm_trampoline!(Actor);
 
-    // Insert the return data block if necessary, and return the correct
-    // block ID.
-    match ret {
-        Ok(None) => NO_DATA_BLOCK_ID,
-        Ok(Some(v)) => match sdk::ipld::put_block(DAG_CBOR, v.bytes()) {
-            Ok(id) => id,
-            Err(e) => abort!(USR_SERIALIZATION, "failed to store return value: {}", e),
-        },
-        Err(e) => abort!(USR_ILLEGAL_STATE, "error calling method: {}", e),
-    }
+/// Atomic execution coordinator actor methods available
+#[derive(FromPrimitive)]
+#[repr(u64)]
+pub enum Method {
+    Constructor = METHOD_CONSTRUCTOR,
+    Join = 2,
+    // Leave = 3,
+    // Kill = 4,
+    // SubmitCheckpoint = 5,
 }
 
 /// SubnetActor trait. Custom subnet actors need to implement this trait
 /// in order to be used as part of hierarchical consensus.
 ///
 /// Subnet actors are responsible for the governing policies of HC subnets.
-pub trait SubnetActor {
```

Comment:
We should keep this interface. This is the interface that every `SubnetActor` should implement to qualify as one (and this is what we are going to seek for to identify the type of the actor as per https://github.com/consensus-shipyard/ipc-gateway/issues/2